### PR TITLE
Support JSON columns in compiled models

### LINKED ISSUE #29602: Compiled models with JSON columns
## Does the Compiled model support JSON columns? 
I did not find any info about this limitation, but when I try to compile a model for context that has configured JSON columns -
```
System.InvalidOperationException: Cannot scaffold C# literals of type 'Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerJsonTypeMapping'. The provider should implement CoreTypeMapping.GenerateCodeLiteral to support using it at design time.
   at Microsoft.EntityFrameworkCore.Design.Internal.CSharpHelper.UnknownLiteral(Object value)
   at Microsoft.EntityFrameworkCore.Design.Internal.CSharpRuntimeAnnotationCodeGenerator.GenerateSimpleAnnotations(CSharpRuntimeAnnotationCodeGeneratorParameters parameters)        
   at Microsoft.EntityFrameworkCore.Design.Internal.CSharpRuntimeAnnotationCodeGenerator.Generate(IEntityType entityType, CSharpRuntimeAnnotationCodeGeneratorParameters parameters) 
   at Microsoft.EntityFrameworkCore.Design.Internal.RelationalCSharpRuntimeAnnotationCodeGenerator.Generate(IEntityType entityType, CSharpRuntimeAnnotationCodeGeneratorParameters parameters)
   at Microsoft.EntityFrameworkCore.SqlServer.Design.Internal.SqlServerCSharpRuntimeAnnotationCodeGenerator.Generate(IEntityType entityType, CSharpRuntimeAnnotationCodeGeneratorParameters parameters)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.CSharpRuntimeModelCodeGenerator.CreateAnnotations[TAnnotatable](TAnnotatable annotatable, Action`2 process, CSharpRuntimeAnnotationCodeGeneratorParameters parameters)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.CSharpRuntimeModelCodeGenerator.CreateAnnotations(IEntityType entityType, IndentedStringBuilder mainBuilder, IndentedStringBuilder methodBuilder, SortedSet`1 namespaces, String className, Boolean nullable)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.CSharpRuntimeModelCodeGenerator.GenerateEntityType(IEntityType entityType, String namespace, String className, Boolean nullable)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.CSharpRuntimeModelCodeGenerator.GenerateModel(IModel model, CompiledModelCodeGenerationOptions options)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.CompiledModelScaffolder.ScaffoldModel(IModel model, String outputDir, CompiledModelCodeGenerationOptions options)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.Optimize(String outputDir, String modelNamespace, String contextTypeName)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OptimizeContextImpl(String outputDir, String modelNamespace, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OptimizeContext.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Cannot scaffold C# literals of type 'Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerJsonTypeMapping'. The provider should implement CoreTypeMapping.GenerateCodeLiteral to support using it at design time.
```

```C#
using Microsoft.EntityFrameworkCore;

public class Info
{
    public string? Code { get; set; }
}

public class Address
{
    public string? Name { get; set; }

    public Info? Info { get; set; }
}

public class Package
{
    public int Id { get; set; }

    public string? Name { get; set; }


    public Address[]? Addresses { get; set; }
}

public class AppContext : DbContext
{

    public DbSet<Package> Packages => Set<Package>();


    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseSqlServer("Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=test123;");
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Package>(o =>
        {
            o.HasKey(x => x.Id);
            o.OwnsMany(w => w.Addresses, b =>
            {
                b.ToJson();
                b.OwnsOne(w => w.Info);
            });
        });

        base.OnModelCreating(modelBuilder);
    }
}

```

EF Core version:
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
Target framework: (e.g. .NET 7.0)
Operating system:
IDE: (e.g. Visual Studio 2022 17.4)

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #29602

The main change here is to store the JSON type mapping in the relational model, and obsolete the previous storage in the underlying EF model.

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs [hunk 1]
```diff
@@ -1321,12 +1322,17 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
 
                 if (index == 0)
                 {
+                    var jsonColumnName = entityType.GetContainerColumnName()!;
+                    var jsonColumnTypeMapping = (entityType.GetViewOrTableMappings().SingleOrDefault()?.Table
+                            ?? entityType.GetDefaultMappings().Single().Table)
+                        .FindColumn(jsonColumnName)!.StoreTypeMapping;
```
Comment (ajcvickers) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1328]: @AndriySvyryd I'm not sure if this is right. Should it just look on the DefaultMappings? Does the column need to be added to both the table mappings and the default mappings? If so, what's the appropriate way to do this? Should it be the same column instance?

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 2]
```diff
@@ -504,7 +504,9 @@ internal SelectExpression(IEntityType entityType, TableExpressionBase tableExpre
         {
             var targetEntityType = ownedJsonNavigation.TargetEntityType;
             var jsonColumnName = targetEntityType.GetContainerColumnName()!;
-            var jsonColumnTypeMapping = targetEntityType.GetContainerColumnTypeMapping()!;
+            var jsonColumnTypeMapping = (entityType.GetViewOrTableMappings().SingleOrDefault()?.Table
+                    ?? entityType.GetDefaultMappings().Single().Table)
+                .FindColumn(jsonColumnName)!.StoreTypeMapping;
```
Comment (ajcvickers) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:509]: Same question here.

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1328]: Default mapping is currently used for `FromSql`, so yes, the column should be added to the default mapping also

Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1328]: Here you should decide whether to use Table, View or default mapping based on what mapping is being used for the container entity type

**Review by AndriySvyryd — APPROVED** [review]

**Comment by ajcvickers** [comment]
@maumar @AndriySvyryd As I understand it, query won't use the default mapping unless there is no table or view mapping. Do we have existing query tests for the default mapping? (Presumably, FromSql tests where the entity type is explicitly not mapped to a table or view?) If so, where?

**Comment by ajcvickers** [comment]
@AndriySvyryd @maumar Updated to store the container column in the default mapping.

**Comment by AndriySvyryd** [comment]
> As I understand it, query won't use the default mapping unless there is no table or view mapping

`FromSql` should always use it, though there might be bugs where a different mapping is used when translating properties than the one used for the entity type.

File: src/EFCore.Relational/Metadata/Internal/ColumnBase.cs [hunk 3]
```diff
@@ -97,7 +99,7 @@ public virtual Type ProviderClrType
     ///     doing so can result in application failures when updating to a new Entity Framework Core release.
     /// </summary>
     public virtual RelationalTypeMapping StoreTypeMapping
-        => PropertyMappings.First().TypeMapping;
+        => _storeTypeMapping ?? PropertyMappings.First().TypeMapping;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Internal/ColumnBase.cs:102]: Store the result in `_storeTypeMapping`

**Review by AndriySvyryd — APPROVED** [review]

**Pull request merged**